internal/controller/promotion/git.go

```diff
@@ -87,73 +87,113 @@ func (g *gitMechanism) GetName() string {
+		var otherStatus *kargoapi.PromotionStatus
+		if otherStatus, newFreight, err = g.doSingleUpdateFn(
 			ctx,
-			stage.Namespace,
+			promo,
 			update,
 			newFreight,
 		); err != nil {
-			return newFreight, err
+			return nil, newFreight, err
 		}
+		newStatus = aggregateGitPromoStatus(newStatus, *otherStatus)
 	}
 
 	logger.Debugf("done executing %s", g.name)
 
-	return newFreight, nil
+	return newStatus, newFreight, nil
 }
 
-// doSingleUpdate updates configuration in a single Git repository.
+// doSingleUpdate updates configuration in a single Git repository by
+// making a git commit with the changes. If performing a pull request
+// promotion, will create a with PR for the git commit instead of
+// committing directly.
 func (g *gitMechanism) doSingleUpdate(
 	ctx context.Context,
-	namespace string,
+	promo *kargoapi.Promotion,
 	update kargoapi.GitRepoUpdate,
 	newFreight kargoapi.SimpleFreight,
-) (kargoapi.SimpleFreight, error) {
+) (*kargoapi.PromotionStatus, kargoapi.SimpleFreight, error) {
 	readRef, commitIndex, err := g.getReadRefFn(update, newFreight.Commits)
 	if err != nil {
-		return newFreight, err
+		return nil, newFreight, err
 	}
 
 	creds, err := g.getCredentialsFn(
 		ctx,
-		namespace,
+		promo.Namespace,
 		update.RepoURL,
 	)
 	if err != nil {
-		return newFreight, err
+		return nil, newFreight, err
+	}
+	if creds == nil {
+		creds = &git.RepoCredentials{}
+	}
+	repo, err := git.Clone(update.RepoURL, *creds, nil)
+	if err != nil {
+		return nil, newFreight, errors.Wrapf(err, "error cloning git repo %q", update.RepoURL)
+	}
+	defer repo.Close()
+
+	commitBranch := update.WriteBranch
+	if update.PullRequest != nil {
+		// When doing a PR promotion, instead of committing to writeBranch directly,
+		// we commit to a temporary, PR branch, which is a child of writeBranch.
+		commitBranch = pullRequestBranchName(promo.Namespace, promo.Spec.Stage)
+
+		if getPullRequestNumberFromMetadata(promo.Status.Metadata, update.RepoURL) == -1 {
+			// PR was never created. Prepare the branch for the commit
+			if err := preparePullRequestBranch(repo, commitBranch, update.WriteBranch); err != nil {
+				return nil, newFreight, errors.Wrapf(err, "error preparing PR branch %q", update.RepoURL)
+			}
+		}
 	}
 
 	commitID, err := g.gitCommitFn(
 		update,
 		newFreight,
 		readRef,
-		update.WriteBranch,
-		creds,
+		commitBranch,
+		repo,
 	)
 	if err != nil {
-		return newFreight, err
+		return nil, newFreight, err
+	}
+
+	newStatus := promo.Status.DeepCopy()
+	if update.PullRequest != nil {
+		newStatus, err = reconcilePullRequest(ctx, promo.Status, repo, creds, commitBranch, update.WriteBranch)
+		if err != nil {
+			return nil, newFreight, err
+		}
+	} else {
+		// For git commit promotions, promotion is successful as soon as the commit is pushed.
+		newStatus.Phase = kargoapi.PromotionPhaseSucceeded
 	}
 
 	if commitIndex > -1 {
 		newFreight.Commits[commitIndex].HealthCheckCommit = commitID
 	}
```

Comment:
There may be another change needed here. If I am reading all new logic above this correctly, in the case of a PR, commitID remains set to the commit at the tip of the PR branch. If `reconcilePullRequest` determines that the PR has been merged, it is the commit now at the tip of the target (stage-specific) branch that should be the new `HealthCheckCommit`. (i.e. The commit that the `Stage` reconciler will be looking for Argo CD Apps to be synced to.)

Reply:
Out of the whole PR, this would be my biggest concern.

Reply:
Great catch. It is a bug.